File: src/driver-app/lib/services/amap_service.dart
import "package:amap_flutter_base/amap_flutter_base.dart";
import "package:dio/dio.dart";
import "package:ride_hermes_driver/config/app_config.dart";

class AmapService {
  final Dio _dio = Dio(BaseOptions(
    connectTimeout: const Duration(seconds: 5),
    receiveTimeout: const Duration(seconds: 5),
  ));

  String get _apiKey => AppConfig.amapApiKey;

  Future<LatLng?> geocode(String address, {String? city}) async {
    if (address.trim().isEmpty) return null;
    try {
      final params = <String, dynamic>{
        "key": _apiKey,
        "address": address,
        "output": "JSON",
      };
      if (city != null && city.isNotEmpty) {
        params["city"] = city;
      }
      final resp = await _dio.get(
        "https://restapi.amap.com/v3/geocode/geo",
        queryParameters: params,
      );
      final data = resp.data as Map<String, dynamic>;
      if (data["status"] == "1") {
        final codes = data["geocodes"] as List?;
        if (codes != null && codes.isNotEmpty) {
          final loc = codes[0]["location"] as String?;
          if (loc != null && loc.contains(",")) {
            final parts = loc.split(",");
            return LatLng(double.parse(parts[1]), double.parse(parts[0]));
          }
        }
      }
    } catch (_) {}
    return null;
  }

  Future<String?> reverseGeocode(double lat, double lng) async {
    try {
      final resp = await _dio.get(
        "https://restapi.amap.com/v3/geocode/regeo",
        queryParameters: {
          "key": _apiKey,
          "location": "$lng,$lat",
          "output": "JSON",
        },
      );
      final data = resp.data as Map<String, dynamic>;
      if (data["status"] == "1") {
        final regeo = data["regeocode"] as Map<String, dynamic>?;
        return regeo?["formatted_address"] as String?;
      }
    } catch (_) {}
    return null;
  }

  Future<List<LatLng>?> getDrivingRoute(LatLng origin, LatLng dest) async {
    try {
      final resp = await _dio.get(
        "https://restapi.amap.com/v3/direction/driving",
        queryParameters: {
          "key": _apiKey,
          "origin": "${origin.longitude},${origin.latitude}",
          "destination": "${dest.longitude},${dest.latitude}",
          "output": "JSON",
        },
      );
      final data = resp.data as Map<String, dynamic>;
      if (data["status"] == "1") {
        final route = data["route"] as Map<String, dynamic>?;
        final paths = route?["paths"] as List?;
        if (paths != null && paths.isNotEmpty) {
          final steps = paths[0]["steps"] as List;
          final points = <LatLng>[];
          for (final step in steps) {
            final polyline = step["polyline"] as String;
            for (final coord in polyline.split(";")) {
              final parts = coord.split(",");
              if (parts.length == 2) {
                points.add(LatLng(
                  double.parse(parts[1]),
                  double.parse(parts[0]),
                ));
              }
            }
          }
          return points;
        }
      }
    } catch (_) {}
    return null;
  }

  Future<String?> getCityFromLocation(double lat, double lng) async {
    try {
      final resp = await _dio.get(
        "https://restapi.amap.com/v3/geocode/regeo",
        queryParameters: {
          "key": _apiKey,
          "location": "$lng,$lat",
          "output": "JSON",
        },
      );
      final data = resp.data as Map<String, dynamic>;
      if (data["status"] == "1") {
        final regeo = data["regeocode"] as Map<String, dynamic>?;
        final comp = regeo?["addressComponent"] as Map<String, dynamic>?;
        if (comp != null) {
          final city = comp["city"];
          if (city is String && city.isNotEmpty) return city;
          final province = comp["province"];
          if (province is String && province.isNotEmpty) return province;
        }
      }
    } catch (_) {}
    return null;
  }

  // === v2.0 新增方法 ===

  /// POI搜索 - 关键词搜索地点
  Future<List<Map<String, dynamic>>> poiSearch(
    String keywords, {
    String? city,
    int limit = 20,
  }) async {
    try {
      final params = <String, dynamic>{
        "key": _apiKey,
        "keywords": keywords,
        "output": "JSON",
        "offset": limit,
        "extensions": "all",
      };
      if (city != null && city.isNotEmpty) {
        params["city"] = city;
      }
      final resp = await _dio.get(
        "https://restapi.amap.com/v3/place/text",
        queryParameters: params,
      );
      final data = resp.data as Map<String, dynamic>;
      if (data["status"] == "1") {
        final pois = data["pois"] as List? ?? [];
        return pois.map<Map<String, dynamic>>((poi) => {
          "name": poi["name"] ?? "",
          "address": poi["address"] ?? "",
          "location": poi["location"] ?? "",
          "tel": poi["tel"] ?? "",
          "type": poi["type"] ?? "",
        }).toList();
      }
    } catch (_) {}
    return [];
  }

  /// 输入提示 - 自动补全
  Future<List<Map<String, dynamic>>> placeSuggestion(
    String keywords, {
    String? city,
  }) async {
    try {
      final params = <String, dynamic>{
        "key": _apiKey,
        "keywords": keywords,
        "output": "JSON",
      };
      if (city != null && city.isNotEmpty) {
        params["city"] = city;
      }
      final resp = await _dio.get(
        "https://restapi.amap.com/v3/assistant/inputtips",
        queryParameters: params,
      );
      final data = resp.data as Map<String, dynamic>;
      if (data["status"] == "1") {
        final tips = data["tips"] as List? ?? [];
        return tips.where((tip) => tip["id"] != null && tip["id"] != "").map<Map<String, dynamic>>((tip) => {
          "name": tip["name"] ?? "",
          "address": tip["address"] ?? "",
          "location": tip["location"] ?? "",
          "district": tip["district"] ?? "",
        }).toList();
      }
    } catch (_) {}
    return [];
  }

  /// 周边搜索 - 搜索附近地点
  Future<List<Map<String, dynamic>>> placeAround(
    double lat, double lng, {
    String keywords = "",
    int radius = 1000,
    int limit = 20,
  }) async {
    try {
      final params = <String, dynamic>{
        "key": _apiKey,
        "location": "$lng,$lat",
        "radius": radius,
        "output": "JSON",
        "offset": limit,
        "extensions": "all",
      };
      if (keywords.isNotEmpty) {
        params["keywords"] = keywords;
      }
      final resp = await _dio.get(
        "https://restapi.amap.com/v3/place/around",
        queryParameters: params,
      );
      final data = resp.data as Map<String, dynamic>;
      if (data["status"] == "1") {
        final pois = data["pois"] as List? ?? [];
        return pois.map<Map<String, dynamic>>((poi) => {
          "name": poi["name"] ?? "",
          "address": poi["address"] ?? "",
          "location": poi["location"] ?? "",
          "distance": poi["distance"] ?? 0,
          "type": poi["type"] ?? "",
        }).toList();
      }
    } catch (_) {}
    return [];
  }

  /// 静态地图URL生成
  String getStaticMapUrl(
    double lat, double lng, {
    int width = 400,
    int height = 300,
    int zoom = 14,
    String markers = "",
  }) {
    final params = {
      "key": _apiKey,
      "center": "$lng,$lat",
      "zoom": zoom,
      "size": "${width}x$height",
    };
    if (markers.isNotEmpty) {
      params["markers"] = markers;
    }
    final query = params.entries.map((e) => "${e.key}=${Uri.encodeComponent(e.value.toString())}").join("&");
    return "https://restapi.amap.com/v3/staticmap?$query";
  }
}

File: src/ride-hermes/internal/service/amap_service.go
package service

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

type AmapService struct {
	apiKey string
	client *http.Client
}

type GeocodeResult struct {
	Address  string  `json:"address"`
	Lat      float64 `json:"lat"`
	Lng      float64 `json:"lng"`
	Resolved bool    `json:"resolved"`
}

func NewAmapService(apiKey string) *AmapService {
	return &AmapService{
		apiKey: apiKey,
		client: &http.Client{Timeout: 10 * time.Second},
	}
}

func (s *AmapService) Geocode(address string) (*GeocodeResult, error) {
	if s.apiKey == "" {
		return &GeocodeResult{Address: address, Resolved: false}, nil
	}

	params := url.Values{}
	params.Set("key", s.apiKey)
	params.Set("address", address)
	params.Set("output", "JSON")

	resp, err := s.client.Get(fmt.Sprintf("https://restapi.amap.com/v3/geocode/geo?%s", params.Encode()))
	if err != nil {
		return &GeocodeResult{Address: address, Resolved: false}, fmt.Errorf("地理编码请求失败: %w", err)
	}
	defer resp.Body.Close()

	var result struct {
		Status   string `json:"status"`
		Geocodes []struct {
			Location string `json:"location"`
		} `json:"geocodes"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return &GeocodeResult{Address: address, Resolved: false}, nil
	}

	if result.Status != "1" || len(result.Geocodes) == 0 {
		return &GeocodeResult{Address: address, Resolved: false}, nil
	}

	location := result.Geocodes[0].Location
	var lng, lat float64
	if _, err := fmt.Sscanf(location, "%f,%f", &lng, &lat); err != nil {
		return &GeocodeResult{Address: address, Resolved: false}, nil
	}

	return &GeocodeResult{
		Address:  address,
		Lat:      lat,
		Lng:      lng,
		Resolved: true,
	}, nil
}

// === v2.0 新增方法 ===

// POI POI搜索结果
type POI struct {
	Name     string `json:"name"`
	Address  string `json:"address"`
	Location string `json:"location"`
	Tel      string `json:"tel,omitempty"`
	Type     string `json:"type,omitempty"`
	Distance string `json:"distance,omitempty"`
}

// POISearch 关键词搜索POI
func (s *AmapService) POISearch(keywords, city string) ([]POI, error) {
	if s.apiKey == "" {
		return nil, fmt.Errorf("AMap API key not configured")
	}

	params := url.Values{}
	params.Set("key", s.apiKey)
	params.Set("keywords", keywords)
	params.Set("output", "JSON")
	params.Set("extensions", "all")
	params.Set("offset", "20")
	if city != "" {
		params.Set("city", city)
	}

	resp, err := s.client.Get(fmt.Sprintf("https://restapi.amap.com/v3/place/text?%s", params.Encode()))
	if err != nil {
		return nil, fmt.Errorf("POI搜索请求失败: %w", err)
	}
	defer resp.Body.Close()

	var result struct {
		Status string `json:"status"`
		POIs   []struct {
			Name     string `json:"name"`
			Address  string `json:"address"`
			Location string `json:"location"`
			Tel      string `json:"tel"`
			Type     string `json:"type"`
		} `json:"pois"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	if result.Status != "1" {
		return []POI{}, nil
	}

	pois := make([]POI, len(result.POIs))
	for i, p := range result.POIs {
		pois[i] = POI{
			Name:     p.Name,
			Address:  p.Address,
			Location: p.Location,
			Tel:      p.Tel,
			Type:     p.Type,
		}
	}
	return pois, nil
}

// PlaceAround 周边搜索POI
func (s *AmapService) PlaceAround(location, keywords, radius string) ([]POI, error) {
	if s.apiKey == "" {
		return nil, fmt.Errorf("AMap API key not configured")
	}

	params := url.Values{}
	params.Set("key", s.apiKey)
	params.Set("location", location)
	params.Set("radius", radius)
	params.Set("output", "JSON")
	params.Set("extensions", "all")
	params.Set("offset", "20")
	if keywords != "" {
		params.Set("keywords", keywords)
	}

	resp, err := s.client.Get(fmt.Sprintf("https://restapi.amap.com/v3/place/around?%s", params.Encode()))
	if err != nil {
		return nil, fmt.Errorf("周边搜索请求失败: %w", err)
	}
	defer resp.Body.Close()

	var result struct {
		Status string `json:"status"`
		POIs   []struct {
			Name     string `json:"name"`
			Address  string `json:"address"`
			Location string `json:"location"`
			Distance string `json:"distance"`
			Type     string `json:"type"`
		} `json:"pois"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	if result.Status != "1" {
		return []POI{}, nil
	}

	pois := make([]POI, len(result.POIs))
	for i, p := range result.POIs {
		pois[i] = POI{
			Name:     p.Name,
			Address:  p.Address,
			Location: p.Location,
			Distance: p.Distance,
			Type:     p.Type,
		}
	}
	return pois, nil
}

// Suggestion 输入提示结果
type Suggestion struct {
	Name     string `json:"name"`
	Address  string `json:"address"`
	Location string `json:"location"`
	District string `json:"district"`
}

// PlaceSuggestion 输入提示/自动补全
func (s *AmapService) PlaceSuggestion(keywords, city string) ([]Suggestion, error) {
	if s.apiKey == "" {
		return nil, fmt.Errorf("AMap API key not configured")
	}

	params := url.Values{}
	params.Set("key", s.apiKey)
	params.Set("keywords", keywords)
	params.Set("output", "JSON")
	if city != "" {
		params.Set("city", city)
	}

	resp, err := s.client.Get(fmt.Sprintf("https://restapi.amap.com/v3/assistant/inputtips?%s", params.Encode()))
	if err != nil {
		return nil, fmt.Errorf("输入提示请求失败: %w", err)
	}
	defer resp.Body.Close()

	var result struct {
		Status string `json:"status"`
		Tips   []map[string]interface{} `json:"tips"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	if result.Status != "1" {
		return []Suggestion{}, nil
	}

	suggestions := make([]Suggestion, 0)
	for _, t := range result.Tips {
		// id 可能是字符串或数组，只处理有有效id的
		idVal := t["id"]
		var idStr string
		switch v := idVal.(type) {
		case string:
			idStr = v
		case []interface{}:
			if len(v) > 0 {
				if s, ok := v[0].(string); ok {
					idStr = s
				}
			}
		}
		if idStr == "" {
			continue
		}
		
		name, _ := t["name"].(string)
		address, _ := t["address"].(string)
		location, _ := t["location"].(string)
		district, _ := t["district"].(string)
		
		suggestions = append(suggestions, Suggestion{
			Name:     name,
			Address:  address,
			Location: location,
			District: district,
		})
	}
	return suggestions, nil
}

// StaticMapURL 生成静态地图URL
func (s *AmapService) StaticMapURL(center, zoom, size, markers string) string {
	params := url.Values{}
	params.Set("key", s.apiKey)
	params.Set("center", center)
	params.Set("zoom", zoom)
	params.Set("size", size)
	if markers != "" {
		params.Set("markers", markers)
	}
	return fmt.Sprintf("https://restapi.amap.com/v3/staticmap?%s", params.Encode())
}

// RouteStep 路线步骤
type RouteStep struct {
	Instruction string `json:"instruction"`
	Road        string `json:"road"`
	Distance    string `json:"distance"`
	Duration    string `json:"duration"`
	Polyline    string `json:"polyline"`
}

// RouteResult 路线规划结果
type RouteResult struct {
	Origin      string      `json:"origin"`
	Destination string      `json:"destination"`
	Distance    string      `json:"distance"`
	Duration    string      `json:"duration"`
	Steps       []RouteStep `json:"steps"`
}

// RoutePlan 路线规划
func (s *AmapService) RoutePlan(origin, destination, mode string) (*RouteResult, error) {
	if s.apiKey == "" {
		return nil, fmt.Errorf("AMap API key not configured")
	}

	params := url.Values{}
	params.Set("key", s.apiKey)
	params.Set("origin", origin)
	params.Set("destination", destination)
	params.Set("output", "JSON")

	var apiURL string
	switch mode {
	case "walking":
		apiURL = fmt.Sprintf("https://restapi.amap.com/v3/direction/walking?%s", params.Encode())
	case "bicycling":
		apiURL = fmt.Sprintf("https://restapi.amap.com/v4/direction/bicycling?%s", params.Encode())
	case "transit":
		params.Set("city", "北京")
		apiURL = fmt.Sprintf("https://restapi.amap.com/v3/direction/transit/integrated?%s", params.Encode())
	default:
		apiURL = fmt.Sprintf("https://restapi.amap.com/v3/direction/driving?%s", params.Encode())
	}

	resp, err := s.client.Get(apiURL)
	if err != nil {
		return nil, fmt.Errorf("路线规划请求失败: %w", err)
	}
	defer resp.Body.Close()

	var result struct {
		Status string `json:"status"`
		Route  struct {
			Origin      string `json:"origin"`
			Destination string `json:"destination"`
			Paths       []struct {
				Distance string `json:"distance"`
				Duration string `json:"duration"`
				Steps    []struct {
					Instruction string `json:"instruction"`
					Road        string `json:"road"`
					Distance    string `json:"distance"`
					Duration    string `json:"duration"`
					Polyline    string `json:"polyline"`
				} `json:"steps"`
			} `json:"paths"`
		} `json:"route"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	if result.Status != "1" || len(result.Route.Paths) == 0 {
		return &RouteResult{
			Origin:      origin,
			Destination: destination,
		}, nil
	}

	path := result.Route.Paths[0]
	steps := make([]RouteStep, len(path.Steps))
	for i, step := range path.Steps {
		steps[i] = RouteStep{
			Instruction: step.Instruction,
			Road:        step.Road,
			Distance:    step.Distance,
			Duration:    step.Duration,
			Polyline:    step.Polyline,
		}
	}

	return &RouteResult{
		Origin:      result.Route.Origin,
		Destination: result.Route.Destination,
		Distance:    path.Distance,
		Duration:    path.Duration,
		Steps:       steps,
	}, nil
}
